### DISCUSSION TITLE
Getting the value of an async method in a non-async path operation function

### DISCUSSION THREAD

**Discussion by orientalperil** [fastapi/fastapi#7352] [Questions] [answered]
I'm using SQLAlchemy ORM so I made my path operation function using `def` not `async def`.  In my function though I want to call `request.json()` which is async and must be awaited.  How can I get the full body of the request in a non-async function?

**Comment by SnkSynthesis** [comment]
@orientalperil 

To call  `request.json()` in synchronous function

You can use `asyncio`

```python
from fastapi import FastAPI, Request
import asyncio

app = FastAPI()

@app.post('/')
def sync_func(request: Request):
    data = asyncio.run(request.json())
    return data
```

or you can use `asgiref` (`pip3 install asgiref`)

```python
from fastapi import FastAPI, Request
from asgiref.sync import async_to_sync

app  = FastAPI()

@app.post('/')
def sync_func(request: Request):
    get_json = async_to_sync(request.json)
    data = get_json()
    return data
```

**Comment by orientalperil** [comment]
Oh right this is so obvious once I see it

**Comment by tiangolo** [comment]
Thanks for the help here @SnkSynthesis! :clap: :bow: 

Thanks for reporting back and closing the issue @orientalperil  :+1:

**Comment by orientalperil** [comment, marked as answer]
Update: This solution is actually not correct.  It works when the size of the request body is small, but will fail if it is large.  You should not run `request.body()` in an async loop other than the main loop.  The correct solution is to call `request.body()` as a dependency before the synchronous request handler code.  An example is at https://github.com/tiangolo/fastapi/issues/2574#issuecomment-751787209
